elastic/elastic-agent
# Pass the parsed version to all downloaders

## What does this PR do?

Avoid stripping build metadata off the version string for agent upgrade versions


## Why is it important?

It should be possible to use build metadata to identify specific builds of a given version for any downloader (fs, http and snapshot)



## Checklist




## Author's Checklist



## How to test this PR locally

Let's create 2 packages with versions containing build metadata from the elastic-agent code:

Note: the way we create the version containing `+buildYYYYMMDD` is a hack and it's not compatible with `SNAPSHOT` versions
- Create the first package camouflaging the agent as an `8.11.1` build by running:
  ```shell
  AGENT_PACKAGE_VERSION=8.11.1+build20231130 BEAT_VERSION=8.11.1  DEV=true EXTERNAL=true PLATFORMS="linux/amd64" PACKAGES=tar.gz mage package
  ```
- Create an empty commit before creating the second package (to avoid collisions in directory names during upgrade):
  ```shell
  git commit --allow-empty -m "Empty commit to change agent hash"
  ```
- Create a second package:
   ```shell
  AGENT_PACKAGE_VERSION=8.11.1+build20231201 BEAT_VERSION=8.11.1  DEV=true EXTERNAL=true PLATFORMS="linux/amd64" PACKAGES=tar.gz mage package
  ```
- Install the first package (no need to enroll in fleet)
- Test that upgrading using the fs downloader works (adapt the `file://` url to the location where the second package is located on disk):
  ```shell
  elastic-agent upgrade 8.11.1+build20231201 --skip-verify --source-uri file:///vagrant/build/distributions/
  ```
- Verify that the upgrade proceeds and that at the end we get an elastic-agent version similar to this:
  ```shell
  elastic-agent version --yaml
  ```
  ```shell
  root@elastic-agent-dev:~# elastic-agent version --yaml
  binary:
    version: 8.11.1+build20231201
    commit: 29def7ee77ea9f10de7cb0b24e85153f07372ec2
    build_time: 2023-12-01T09:59:54Z
    snapshot: false
  daemon:
    version: 8.11.1+build20231201
    commit: 29def7ee77ea9f10de7cb0b24e85153f07372ec2
    build_time: 2023-12-01T09:59:54Z
    snapshot: false
  ```
- Create a simple directory structure, link the last part of the path to where the packages are located and start a simple http server (in the example I am using `python`)
  ```shell
  mkdir beats
  ln -s /vagrant/build/distributions beats/elastic-agent
   python3 -m http.server 
  ```
  when the http server starts, it should look like this
  ```shell
  vagrant@elastic-agent-dev:~$  python3 -m http.server
  Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...
  ```
- Now we can test the http downloader by specifying a `source-uri` by going back to the previous version
  ```shell
    root@elastic-agent-dev:~# elastic-agent upgrade 8.11.1+build20231130 --skip-verify --source-uri http://localhost:8000
  Upgrade triggered to version 8.11.1+build20231130, Elastic Agent is currently restarting
  ```
  ```shell
  root@elastic-agent-dev:~# elastic-agent version --yaml                                                    
  binary:
    version: 8.11.1+build20231130
    commit: 8439fe6a814319e3ad9575abd5dcc01012678d35
    build_time: 2023-12-01T13:50:11Z
    snapshot: false
  daemon:
    version: 8.11.1+build20231130
    commit: 8439fe6a814319e3ad9575abd5dcc01012678d35
    build_time: 2023-12-01T13:50:11Z
    snapshot: false

  ```



## Related issues


- Closes #3813

## Use cases



## Screenshots



## Logs



## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
